src/models/base_model.py

```diff
@@ -0,0 +1,226 @@
+            hook_name = hook.module_name
+            module = modules[hook_name]
+            hook_fn = partial(self._collect_output_hook, hook_name)
+            # registrate hooks
+            if hook.hook_type == HookType.FORWARD_PRE:
+                module.register_forward_pre_hook(hook_fn)
+            else:
+                module.register_forward_hook(hook_fn)
+           
+        self._feature_outputs = defaultdict(OrderedDict)
+
+    def _collect_output_hook(self, hook_name: str, *args):
+        """Collect hooks outputs in self._feature_outputs.
+        
+        Args:
+            hook_name: Name of saved hooked module.
+        """
+        # tensor we want is last argument, output for fwd, input for fwd_pre
+        x = args[-1]  
+        if isinstance(x, (tuple, list)) and len(x) == 1:
+            # unwrap input tuple
+            x = x[0]  
+            device = x.device
+        else:
+            device = x[0].device
+        self._feature_outputs[device][hook_name] = x
+
+    def get_output(self, device: torch.device) -> Dict[str, torch.tensor]:
+        """Return hooks output.
+
+        Args:
+            device: Saved hooks device.
+
+        Returns:
+            output: Dict of module hook name two its output tensor.
+        """
+        output = self._feature_outputs[device]
+        # clear after reading
+        self._feature_outputs[device] = OrderedDict()  
+        return output
+
+
+class BaseModel(nn.Module):
+    """Base model for all Torchok Models - Backbone, Neck, Pooling and Head.
+
+    This class support methods for add hooks.
+    Also it's have channels and reduction for input and output tensors, for each forward and hooks case. 
+    Channels - tensor channels.
+    Reduction -  downscale value for get tensor shape.
+
+    Args:
+        input_forward_channels: Input channels list for forward method.
+        input_forward_reductions: Optional value of input reduction list for forward. Need to reconstruct initial input
+            tensor shape.
+        input_hooks_channels: Optional value of input channels list for hooks.
+        input_hooks_reductions: Optional value of input hooks reduction list. Need to reconstruct initial input tensor 
+            shape.
+    """
+    def __init__(self, input_forward_channels: List[int], input_forward_reductions: List[int] = None, \
+                 input_hooks_channels: List[int] = None, input_hooks_reductions: List[int] = None):
+        super().__init__()
+        self._feature_info = None
+
+        self._input_forward_channels = input_forward_channels
+        self._input_forward_reductions = input_forward_reductions
+        self._input_hooks_channels = input_hooks_channels
+        self._input_hooks_reductions = input_hooks_reductions
+
+        self._output_forward_channels = None
+        self._output_forward_reductions = None
+        self._output_hooks_channels = None
+        self._output_hooks_reductions = None
+
+    def create_hooks(self):
```

Comment:
1. This method has to be called on a model initialization stage. So, it needs to be clearly explained in the docstring. Otherwise, user will not know about this functionality.
2. Move arguments from `__init__` to `create_hooks` because it's related to the same functionality which isn't used by default. Also wrap these parameters into a list of `FeaureInfo`, it will be easier for user to pass